Support reading arrays as List<T>

Brar opened the pull request:
This is an initial PR to implement #1821 
It is mostly for discussion now as I intentionally didn't optimize anything yet.
The List\<T> is created via reflection and initialized with the array that has been read from the database.

IMO there are two options for optimization but I'm not sure if this rare case justifies additional code complexity:

1. Use delegates to create the List\<T>
2. Read the values from the buffer directly into the List\<T>

No 1 would certainly be faster but make the code ways more complicated ([example](https://codeblog.jonskeet.uk/2008/08/09/making-reflection-fly-and-exploring-delegates/)).
No 2 would save the allocation of one copy of the array to return but would lead to a separate implementation of `ArrayHandler.Read<TElement2>` just for this case.

@roji what do you think?

Brar commented:
I've added a Benchmark so that we have a better base for discussion.
If we decide to optimize the two options mentioned above, we should get results that are pretty similar for `AsArray` and `AsListOfT`

### Environment

```
BenchmarkDotNet=v0.10.12, OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.309)
Intel Core i7 CPU 920 2.67GHz (Nehalem), 1 CPU, 8 logical cores and 4 physical cores
Frequency=2604245 Hz, Resolution=383.9884 ns, Timer=TSC
.NET Core SDK=2.1.101
  [Host]     : .NET Core 2.0.6 (Framework 4.6.26212.01), 64bit RyuJIT
  DefaultJob : .NET Core 2.0.6 (Framework 4.6.26212.01), 64bit RyuJIT
```

### ReadArrayOfIntegers

Method | NumArrayElements |            Mean |         Error |          StdDev |          Median |         Op/s |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
---------- |-----------------:|----------------:|--------------:|----------------:|----------------:|-------------:|---------:|---------:|---------:|----------:|
AsArray |                0 |        757.1 ns |      14.64 ns |        16.27 ns |        757.0 ns | 1,320,873.94 |   0.0095 |        - |        - |      48 B |
AsListOfT |                0 |      6,265.1 ns |     145.69 ns |       287.58 ns |      6,255.8 ns |   159,614.12 |   0.1297 |        - |        - |     544 B |
AsArray |               10 |      3,931.3 ns |      34.18 ns |        31.97 ns |      3,931.0 ns |   254,370.82 |   0.0153 |        - |        - |      96 B |
AsListOfT |               10 |     10,099.6 ns |     205.43 ns |       596.00 ns |     10,085.7 ns |    99,013.66 |   0.1526 |        - |        - |     656 B |
AsArray |             1000 |    294,282.8 ns |   9,227.95 ns |    26,918.36 ns |    295,733.9 ns |     3,398.09 |   0.4883 |        - |        - |    4056 B |
AsListOfT |             1000 |    553,319.8 ns |  53,193.05 ns |   156,840.92 ns |    492,850.8 ns |     1,807.27 |   1.9531 |        - |        - |    8579 B |
AsArray |           100000 | 38,371,069.0 ns | 761,585.63 ns | 1,687,621.16 ns | 38,389,019.7 ns |        26.06 |  62.5000 |  62.5000 |  62.5000 |  400056 B |
AsListOfT |           100000 | 38,539,158.3 ns | 770,455.99 ns | 1,674,907.92 ns | 38,447,681.1 ns |        25.95 | 125.0000 | 125.0000 | 125.0000 |  801672 B |

### ReadArrayOfStrings

Method | NumArrayElements |            Mean |           Error |          StdDev |         Op/s |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
---------- |-----------------:|----------------:|----------------:|----------------:|-------------:|---------:|---------:|---------:|----------:|
AsArray |                0 |        872.5 ns |        18.90 ns |        55.14 ns | 1,146,123.97 |   0.0105 |        - |        - |      48 B |
AsListOfT |                0 |      6,696.0 ns |       133.34 ns |       124.73 ns |   149,343.29 |   0.1297 |        - |        - |     544 B |
AsArray |               10 |      5,944.9 ns |        47.19 ns |        31.21 ns |   168,210.42 |   0.1221 |        - |        - |     536 B |
AsListOfT |               10 |     12,555.0 ns |       145.37 ns |       128.87 ns |    79,649.40 |   0.2594 |        - |        - |    1136 B |
AsArray |             1000 |    410,555.8 ns |     4,603.52 ns |     3,328.64 ns |     2,435.72 |  11.2305 |        - |        - |   48056 B |
AsListOfT |             1000 |    426,521.1 ns |     2,365.02 ns |     2,096.53 ns |     2,344.55 |  13.1836 |        - |        - |   56594 B |
AsArray |           100000 | 61,056,514.9 ns | 1,210,201.02 ns | 2,119,569.47 ns |        16.38 | 687.5000 | 312.5000 | 125.0000 | 4800056 B |
AsListOfT |           100000 | 61,566,606.2 ns |   363,121.21 ns |   321,897.46 ns |        16.24 | 812.5000 | 437.5000 | 250.0000 | 5602560 B |

Brar commented:
As nobody seems to have time to come out and play with me, I played a little by myself and came up with a solution that reads directly from the buffer and as an additional benefit doesn't need any reflection or delegates to read the array.
It performs ways better in terms of Op/s and also allocates less memory and by that produces less GC pressure.
The newly introduced `ReadAsList` method is pretty simple as it only handles arrays with dimensions < 2 so the overall complexity added by this PR seems OK for me.

### ReadArrayOfIntegers

Method | NumArrayElements |            Mean |          Error |         StdDev |         Op/s |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
---------- |-----------------:|----------------:|---------------:|---------------:|-------------:|--------:|--------:|--------:|----------:|
AsArray |                0 |        835.1 ns |      20.683 ns |      60.335 ns | 1,197,481.27 |  0.0095 |       - |       - |      48 B |
AsListOfT |                0 |        491.9 ns |       3.402 ns |       2.841 ns | 2,033,055.20 |  0.0086 |       - |       - |      40 B |
AsArray |               10 |      3,888.8 ns |      73.614 ns |     153.659 ns |   257,147.82 |  0.0153 |       - |       - |      96 B |
AsListOfT |               10 |      3,374.7 ns |      33.358 ns |      31.203 ns |   296,318.44 |  0.0229 |       - |       - |     104 B |
AsArray |             1000 |    282,829.8 ns |   5,630.720 ns |   7,893.483 ns |     3,535.70 |  0.4883 |       - |       - |    4056 B |
AsListOfT |             1000 |    293,931.7 ns |   2,428.225 ns |   1,895.798 ns |     3,402.15 |  0.4883 |       - |       - |    4064 B |
AsArray |           100000 | 28,242,794.7 ns | 583,645.289 ns | 545,942.113 ns |        35.41 | 93.7500 | 93.7500 | 93.7500 |  400056 B |
AsListOfT |           100000 | 31,339,837.5 ns | 594,680.858 ns | 636,301.797 ns |        31.91 | 62.5000 | 62.5000 | 62.5000 |  400064 B |


### ReadArrayOfStrings

Method | NumArrayElements |            Mean |           Error |           StdDev |         Op/s |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
---------- |-----------------:|----------------:|----------------:|-----------------:|-------------:|---------:|---------:|---------:|----------:|
AsArray |                0 |        759.3 ns |        10.22 ns |         9.562 ns | 1,316,935.64 |   0.0105 |        - |        - |      48 B |
AsListOfT |                0 |        532.8 ns |        10.78 ns |        20.516 ns | 1,876,817.01 |   0.0086 |        - |        - |      40 B |
AsArray |               10 |      6,609.7 ns |        59.97 ns |        53.164 ns |   151,292.58 |   0.1221 |        - |        - |     536 B |
AsListOfT |               10 |      6,658.2 ns |       140.29 ns |       226.545 ns |   150,190.14 |   0.1221 |        - |        - |     544 B |
AsArray |             1000 |    408,974.0 ns |     3,285.92 ns |     2,743.890 ns |     2,445.14 |  11.2305 |        - |        - |   48056 B |
AsListOfT |             1000 |    456,581.2 ns |     8,056.84 ns |     6,727.830 ns |     2,190.19 |  11.2305 |        - |        - |   48064 B |
AsArray |           100000 | 62,736,933.8 ns | 1,172,543.62 ns | 1,303,279.004 ns |        15.94 | 750.0000 | 375.0000 | 125.0000 | 4800410 B |
AsListOfT |           100000 | 78,877,692.9 ns | 1,562,987.38 ns | 2,696,076.441 ns |        12.68 | 750.0000 | 375.0000 | 125.0000 | 4800247 B |

YohDeadfall commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -91,21 +91,59 @@ public ArrayHandler([CanBeNull] NpgsqlTypeHandler elementHandler, int lowerBound
+            buf.ReadInt32();        // Has nulls. Not populated by PG?
+            var elementOID = buf.ReadUInt32();
+            // The following should hold but fails in test CopyTests.ReadBitString
+            //Debug.Assert(elementOID == ElementHandler.BackendType.OID);
```
There are two bit types in PostgreSQL: `bit` and `varbit`. In the `ReadBitString` test you get the second one `oid` which `oid` is 1562. The element handler should provide multiple `oid` in such cases, but it's a different issue.

YohDeadfall replied:
/cc @roji

Brar replied:
This was a copy-pasted comment from ` Read\<TElement2>`.
I've removed it here.

YohDeadfall commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -91,21 +91,59 @@ public ArrayHandler([CanBeNull] NpgsqlTypeHandler elementHandler, int lowerBound
         }
 
+        protected internal virtual ValueTask<object> ReadPsvAsList(NpgsqlReadBuffer buf, bool async)
+            => throw new NotSupportedException();
```
Remove. See the comment above.

Brar replied:
I don't understand what you mean here. IMO removing this method would mean removing the functionality of reading arrays as List of provider specific values.

YohDeadfall commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -353,6 +391,9 @@ internal override object ReadPsvAsObject(NpgsqlReadBuffer buf, int len, FieldDes
             => await Read<TElementPsv>(buf, async);
 
+        protected internal override ValueTask<object> ReadPsvAsList(NpgsqlReadBuffer buf, bool async)
+            => ReadAsList<TElementPsv>(buf, async);
```
Remove. See the comment above.

Brar replied:
See answer above.

Brar commented on src/Npgsql/TypeHandlers/ArrayHandler.cs:
```diff
@@ -91,21 +91,59 @@ public ArrayHandler([CanBeNull] NpgsqlTypeHandler elementHandler, int lowerBound
+            {
+                return (TArray)await ReadPsvAsList(buf, async);
+            }
+                throw new InvalidCastException($"Can't cast database type {PgDisplayName} to {typeof(TArray).Name}");
```
Fixed.

Brar commented on test/Npgsql.Benchmarks/ReadArray.cs:
```diff
@@ -0,0 +1,76 @@
+    }
+
+
+    public abstract class ReadArray<T>
```
Yes. Thanks for the hint!

Brar commented:
Ok @YohDeadfall, thank you for your review and your useful hints.

I think I've addressed all the comments.
The following is the new Bechmark.
Benchmarks for NpgsqlInet test the provider specific value.

Method | NumArrayElements |             Mean |            Error |           StdDev |           Median |          Op/s |     Gen 0 |     Gen 1 |    Gen 2 |  Allocated |
--------------------- |-----------------:|-----------------:|-----------------:|-----------------:|-----------------:|--------------:|----------:|----------:|---------:|-----------:|
ReadIntArray |                0 |         766.0 ns |        21.796 ns |        63.924 ns |         763.8 ns | 1,305,453.036 |    0.0105 |         - |        - |       48 B |
ReadListOfInt |                0 |       1,222.4 ns |        78.682 ns |       231.995 ns |       1,283.6 ns |   818,084.687 |    0.0086 |         - |        - |       40 B |
ReadStringArray |                0 |         762.8 ns |         8.166 ns |         7.239 ns |         763.0 ns | 1,310,924.187 |    0.0105 |         - |        - |       48 B |
ReadListOfString |                0 |         992.5 ns |       148.003 ns |       436.389 ns |         694.8 ns | 1,007,605.379 |    0.0076 |         - |        - |       40 B |
ReadIPAddressArray |                0 |       1,741.2 ns |       256.641 ns |       756.713 ns |       1,916.1 ns |   574,308.080 |    0.0105 |         - |        - |       48 B |
ReadListOfIPAddress |                0 |         608.1 ns |        12.228 ns |        27.600 ns |         598.8 ns | 1,644,375.581 |    0.0086 |         - |        - |       40 B |
ReadNpgsqlInetArray |                0 |       1,574.4 ns |        79.299 ns |       232.571 ns |       1,551.1 ns |   635,144.402 |    0.0076 |         - |        - |       48 B |
ReadListOfNpgsqlInet |                0 |       1,158.8 ns |        23.202 ns |        64.678 ns |       1,158.8 ns |   862,963.898 |    0.0153 |         - |        - |       72 B |
ReadIntArray |               10 |       3,797.3 ns |        72.319 ns |        64.109 ns |       3,793.5 ns |   263,344.566 |    0.0153 |         - |        - |       96 B |
ReadListOfInt |               10 |       3,416.9 ns |        39.029 ns |        34.598 ns |       3,425.4 ns |   292,662.806 |    0.0229 |         - |        - |      104 B |
ReadStringArray |               10 |       8,591.8 ns |       171.590 ns |       424.129 ns |       8,622.6 ns |   116,389.696 |    0.4425 |         - |        - |     1896 B |
ReadListOfString |               10 |       7,010.0 ns |        78.414 ns |        69.512 ns |       6,995.9 ns |   142,653.158 |    0.4501 |         - |        - |     1904 B |
ReadIPAddressArray |               10 |       7,534.7 ns |       180.118 ns |       522.555 ns |       7,394.4 ns |   132,719.707 |    0.1984 |         - |        - |      856 B |
ReadListOfIPAddress |               10 |       6,615.7 ns |        85.127 ns |        66.461 ns |       6,610.3 ns |   151,155.124 |    0.1984 |         - |        - |      864 B |
ReadNpgsqlInetArray |               10 |       6,627.0 ns |       130.511 ns |       199.303 ns |       6,586.6 ns |   150,897.408 |    0.2213 |         - |        - |      936 B |
ReadListOfNpgsqlInet |               10 |       6,563.6 ns |       133.145 ns |       124.544 ns |       6,526.3 ns |   152,354.277 |    0.2289 |         - |        - |      976 B |
ReadIntArray |             1000 |     278,293.2 ns |     2,719.312 ns |     2,410.599 ns |     277,733.9 ns |     3,593.332 |    0.4883 |         - |        - |     4056 B |
ReadListOfInt |             1000 |     280,362.5 ns |     1,813.892 ns |     1,696.715 ns |     280,622.9 ns |     3,566.811 |    0.4883 |         - |        - |     4064 B |
ReadStringArray |             1000 |     518,769.3 ns |     6,328.201 ns |     5,609.785 ns |     518,523.7 ns |     1,927.639 |   41.9922 |    5.8594 |        - |   184056 B |
ReadListOfString |             1000 |     653,749.7 ns |    10,313.900 ns |     9,647.628 ns |     653,850.1 ns |     1,529.638 |   42.9688 |         - |        - |   184064 B |
ReadIPAddressArray |             1000 |     541,835.2 ns |     3,160.837 ns |     2,467.773 ns |     541,721.6 ns |     1,845.580 |   18.5547 |         - |        - |    80056 B |
ReadListOfIPAddress |             1000 |     580,303.1 ns |     4,738.363 ns |     4,200.435 ns |     579,085.1 ns |     1,723.237 |   18.5547 |         - |        - |    80064 B |
ReadNpgsqlInetArray |             1000 |     496,467.3 ns |     3,911.557 ns |     3,467.493 ns |     495,771.7 ns |     2,014.231 |   20.5078 |    0.9766 |        - |    88056 B |
ReadListOfNpgsqlInet |             1000 |     518,117.1 ns |     5,605.641 ns |     4,969.254 ns |     518,919.8 ns |     1,930.066 |   20.5078 |    0.9766 |        - |    88096 B |
ReadIntArray |           100000 |  28,098,648.9 ns |   555,171.675 ns |   864,335.413 ns |  28,291,817.1 ns |        35.589 |   93.7500 |   93.7500 |  93.7500 |   400056 B |
ReadListOfInt |           100000 |  27,515,868.5 ns |   191,803.410 ns |   179,413.011 ns |  27,534,494.3 ns |        36.343 |   93.7500 |   93.7500 |  93.7500 |   400064 B |
ReadStringArray |           100000 | 104,632,831.7 ns | 1,133,492.708 ns | 1,060,269.679 ns | 104,469,226.5 ns |         9.557 | 3500.0000 | 2062.5000 | 687.5000 | 18400676 B |
ReadListOfString |           100000 | 119,111,465.1 ns | 1,750,294.208 ns | 1,461,574.795 ns | 118,689,024.8 ns |         8.395 | 3375.0000 | 1750.0000 | 562.5000 | 18400760 B |
ReadIPAddressArray |           100000 |  91,657,797.0 ns | 1,726,094.518 ns | 1,530,137.681 ns |  91,223,227.0 ns |        10.910 | 1312.5000 |  625.0000 | 187.5000 |  8000372 B |
ReadListOfIPAddress |           100000 |  96,403,993.1 ns | 1,803,513.199 ns | 1,687,007.201 ns |  96,118,229.6 ns |        10.373 | 1312.5000 |  625.0000 | 187.5000 |  8000266 B |
ReadNpgsqlInetArray |           100000 |  83,657,972.2 ns | 1,212,953.426 ns | 1,134,597.276 ns |  83,246,207.3 ns |        11.953 | 1500.0000 |  875.0000 | 375.0000 |  8800312 B |
ReadListOfNpgsqlInet |           100000 |  85,238,271.6 ns | 1,111,948.328 ns |   985,713.133 ns |  85,285,740.6 ns |        11.732 | 1500.0000 |  875.0000 | 375.0000 |  8801083 B |

roji commented:
Note: squashed post-merge to eb62d0a6ac61c9be351a1f4261ef111115984ddb

YohDeadfall commented:
Okay, will push it tomorrow.
